File: src/scripts/migrate-ig.py
import os
import sys
import re
import random
import string
import requests

def parse_caption(caption):
    lines = caption.split('\n')
    first_line = lines[0].strip() if lines else ""
    
    price = 0
    group = ''
    album_era = ''
    member = ''
    
    # 1. Extract Price
    price_match = re.search(r'\$(\d+)(?:\s+set)?', caption, re.IGNORECASE)
    if price_match:
        price = float(price_match.group(1))
        
    # 2. Extract Group and Album from the first line
    # Format: #hashtag GroupName AlbumName SpecialDetail
    if first_line.startswith('#'):
        tokens = [t.strip() for t in first_line.split() if t.strip()]
        if len(tokens) >= 2:
            group = tokens[1]
            lower_group = group.lower()
            # Beautify known groups
            if lower_group == 'p1harmony':
                group = 'P1Harmony'
            elif lower_group == 'illit':
                group = 'Illit'
            elif lower_group == 'ampers&one':
                group = 'Ampers&one'
            elif lower_group == '&team':
                group = '&Team'
            elif lower_group == 'xikers':
                group = 'Xikers'
            elif lower_group == 'riize':
                group = 'Riize'
                
        if len(tokens) >= 3:
            album_era = tokens[2]
            
    # 3. Extract Title (strip first token if it starts with #)
    title = first_line
    if first_line.startswith('#'):
        tokens = [t.strip() for t in first_line.split() if t.strip()]
        if len(tokens) >= 2:
            title = " ".join(tokens[1:])
    title = title[:80].strip() if title else 'IG Post'
    
    return title, price, group, album_era, member

def main():
    print("🚀 Starting Instagram to Supabase migration using Python (SOCKS5)...")

    # 1. Parse .env.local manually
    env = {}
    env_path = '.env.local'
    if os.path.exists(env_path):
        with open(env_path, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                if '=' in line:
                    key, val = line.split('=', 1)
                    env[key.strip()] = val.strip().strip("'").strip('"')

    supabase_url = env.get('NEXT_PUBLIC_SUPABASE_URL')
    supabase_key = env.get('SUPABASE_SERVICE_ROLE_KEY') or env.get('NEXT_PUBLIC_SUPABASE_ANON_KEY')
    proxy = env.get('proxy')

    if not supabase_url or not supabase_key:
        print("❌ Error: Supabase credentials not found in .env.local!")
        sys.exit(1)

    # Force SOCKS5h to bypass TLS connect proxy handshake errors
    proxies = None
    if proxy:
        clean_proxy = proxy.replace('http://', '').replace('https://', '')
        socks_proxy = f"socks5h://{clean_proxy}"
        print(f"🌐 Using SOCKS5 proxy: {socks_proxy}")
        proxies = {
            "http": socks_proxy,
            "https": socks_proxy
        }

    # Set base request headers
    headers = {
        "apikey": supabase_key,
        "Authorization": f"Bearer {supabase_key}"
    }

    # 2. Fetch existing migrated cards from database
    print("🔍 Fetching existing migrated cards from database...")
    try:
        url = f"{supabase_url}/rest/v1/cards?select=title,description"
        response = requests.get(url, headers=headers, proxies=proxies, timeout=15)
        response.raise_for_status()
        existing_cards = response.json()
    except Exception as e:
        print(f"❌ Failed to fetch existing cards from Supabase: {e}")
        sys.exit(1)

    existing_keys = set(f"{c.get('title')}::{c.get('description')}" for c in existing_cards)
    print(f"ℹ️ Found {len(existing_keys)} existing cards in database.")

    # 3. Scan export directory
    export_dir = 'ig_export'
    if not os.path.exists(export_dir):
        print(f"❌ Folder '{export_dir}' not found! Please create it and run download script first.")
        sys.exit(1)

    files = os.listdir(export_dir)
    txt_files = [f for f in files if f.endswith('.txt')]

    print(f"🚀 Starting migration of {len(txt_files)} posts...")

    success_count = 0
    fail_count = 0
    skipped_count = 0

    for i, txt_file in enumerate(txt_files):
        base_name = txt_file[:-4]
        
        # Find corresponding image
        img_file = None
        for ext in ['.jpg', '_1.jpg', '.png']:
            possible_file = f"{base_name}{ext}"
            if possible_file in files:
                img_file = possible_file
                break
                
        if not img_file:
            print(f"[{i+1}/{len(txt_files)}] ⚠️ No image found for {txt_file}, skipping.")
            continue

        try:
            # Read Caption
            with open(os.path.join(export_dir, txt_file), 'r', encoding='utf-8') as f:
                caption = f.read()
                
            title, price, group, album_era, member = parse_caption(caption)

            # Check if already migrated
            key = f"{title}::{caption}"
            if key in existing_keys:
                skipped_count += 1
                continue

            print(f"🔄 Migrating: {title} (File: {base_name})")

            # Upload Image to Storage Bucket 'cards'
            img_path = os.path.join(export_dir, img_file)
            random_str = ''.join(random.choices(string.ascii_lowercase + string.digits, k=6))
            unique_name = f"{int(os.path.getmtime(img_path))}-{random_str}.jpg"
            
            storage_url = f"{supabase_url}/storage/v1/object/cards/migrated/{unique_name}"
            upload_headers = {
                **headers,
                "Content-Type": "image/jpeg"
            }
            
            with open(img_path, 'rb') as img_data:
                upload_resp = requests.post(storage_url, headers=upload_headers, data=img_data, proxies=proxies, timeout=30)
                upload_resp.raise_for_status()

            # Construct Public URL
            public_url = f"{supabase_url}/storage/v1/object/public/cards/migrated/{unique_name}"

            # Insert into database 'cards' table
            db_url = f"{supabase_url}/rest/v1/cards"
            insert_headers = {
                **headers,
                "Content-Type": "application/json",
                "Prefer": "return=representation"
            }
            payload = {
                "title": title,
                "description": caption,
                "price": price,
                "group_name": group,
                "album_era": album_era,
                "member_name": member,
                "image_url": public_url,
                "source": "instagram",
                "original_ig_url": f"https://www.instagram.com/p/{base_name.split('_')[0]}/"
            }
            
            insert_resp = requests.post(db_url, headers=insert_headers, json=payload, proxies=proxies, timeout=15)
            insert_resp.raise_for_status()

            success_count += 1
            if success_count % 10 == 0:
                print(f"✅ Progress: {success_count}/{len(txt_files)} ({int(success_count/len(txt_files)*100)}%)")

        except Exception as e:
            fail_count += 1
            print(f"❌ [{i+1}/{len(txt_files)}] Failed {base_name}: {e}")

    print("\n🏁 Migration Complete!")
    print(f"📊 Success: {success_count}")
    print(f"⏭️ Skipped (already exist): {skipped_count}")
    print(f"❌ Failed: {fail_count}")

if __name__ == '__main__':
    main()

File: src/scripts/migrate-ig.ts
/**
 * Instagram to Supabase Migration Script
 * 
 * Usage:
 * 1. Put your exported images and .txt files in a folder named 'ig_export'
 * 2. Run: npx ts-node src/scripts/migrate-ig.ts
 */

import * as fs from 'fs';
import * as path from 'path';
import { createClient } from '@supabase/supabase-js';
import * as dotenv from 'dotenv';
import nodeFetch from 'node-fetch';
import { SocksProxyAgent } from 'socks-proxy-agent';

dotenv.config({ path: '.env.local' });

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;
const proxy = process.env.proxy;

let customFetch: typeof nodeFetch = nodeFetch;
if (proxy) {
  const cleanProxy = proxy.replace('http://', '').replace('https://', '');
  const socksProxy = `socks5h://${cleanProxy}`;
  console.log(`🌐 Routing Supabase requests through SOCKS5 proxy: ${socksProxy}`);
  const agent = new SocksProxyAgent(socksProxy);
  customFetch = (url, init) => {
    return nodeFetch(url, {
      ...init,
      agent
    });
  };
}

const supabase = createClient(supabaseUrl, supabaseKey, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  },
  global: {
    fetch: customFetch as unknown as typeof fetch
  }
});

const EXPORT_DIR = path.join(process.cwd(), 'ig_export');

type MigratedCardInsert = {
  title: string;
  description: string;
  price: number;
  group_name: string;
  album_era: string;
  member_name: string;
  image_url: string;
  source: 'instagram';
  original_ig_url: string;
};

async function getExistingStorageFiles(): Promise<Set<string>> {
  const existingFiles = new Set<string>();
  let offset = 0;
  const limit = 1000;
  
  while (true) {
    console.log(`   Fetching storage files (offset: ${offset})...`);
    const { data, error } = await supabase.storage
      .from('cards')
      .list('migrated', {
        limit,
        offset,
        sortBy: { column: 'name', order: 'asc' }
      });
      
    if (error) {
      console.error(`❌ Error listing storage files:`, error.message);
      break;
    }
    
    if (!data || data.length === 0) {
      break;
    }
    
    for (const item of data) {
      if (item.name) {
        existingFiles.add(item.name);
      }
    }
    
    if (data.length < limit) {
      break;
    }
    
    offset += limit;
  }
  
  console.log(`ℹ️ Found ${existingFiles.size} existing files in Supabase Storage.`);
  return existingFiles;
}

async function asyncPool<T, R>(
  concurrency: number,
  iterable: T[],
  iteratorFn: (item: T) => Promise<R>
): Promise<R[]> {
  const results: Promise<R>[] = [];
  const executing: Promise<void>[] = [];
  
  for (const item of iterable) {
    const p = Promise.resolve().then(() => iteratorFn(item));
    results.push(p);
    
    if (concurrency <= iterable.length) {
      const e: Promise<void> = p.then(() => {
        executing.splice(executing.indexOf(e), 1);
      });
      executing.push(e);
      if (executing.length >= concurrency) {
        await Promise.race(executing);
      }
    }
  }
  return Promise.all(results);
}

async function getExistingDbKeys(): Promise<Set<string>> {
  const existingKeys = new Set<string>();
  let offset = 0;
  const limit = 1000;
  
  while (true) {
    console.log(`   Fetching database cards (offset: ${offset})...`);
    const { data, error } = await supabase
      .from('cards')
      .select('original_ig_url')
      .range(offset, offset + limit - 1);
      
    if (error) {
      console.error(`❌ Error fetching cards from database:`, error.message);
      break;
    }
    
    if (!data || data.length === 0) {
      break;
    }
    
    for (const item of data) {
      if (item.original_ig_url) {
        existingKeys.add(item.original_ig_url);
      }
    }
    
    if (data.length < limit) {
      break;
    }
    
    offset += limit;
  }
  
  console.log(`ℹ️ Found ${existingKeys.size} existing Instagram cards in database.`);
  return existingKeys;
}

async function migrate() {
  console.log('🚀 Starting highly optimized bulk migration with complete image check...');

  if (!fs.existsSync(EXPORT_DIR)) {
    console.error(`❌ Folder 'ig_export' not found! Please create it and put your IG data inside.`);
    return;
  }

  const files = fs.readdirSync(EXPORT_DIR);
  const textFiles = files.filter(f => f.endsWith('.txt'));

  console.log(`🔍 Fetching existing storage files to see what is already uploaded...`);
  const existingStorageFiles = await getExistingStorageFiles();

  console.log(`🔍 Fetching existing migrated cards from database...`);
  const existingDbKeys = await getExistingDbKeys();

  console.log(`🚀 Processing metadata for ${textFiles.length} posts...`);

  const cardsToInsert: MigratedCardInsert[] = [];
  const uploadTasks: { baseName: string; fileName: string; imgPath: string }[] = [];
  let alreadyExistInBoth = 0;
  let onlyNeedUpload = 0;

  for (let i = 0; i < textFiles.length; i++) {
    const txtFile = textFiles[i];
    const baseName = txtFile.replace('.txt', '');
    
    // Find the primary image
    const imgFile = files.find(f => 
      (f === `${baseName}.jpg` || f === `${baseName}_1.jpg` || f === `${baseName}.png`)
    );
    
    if (!imgFile) {
      continue;
    }

    const targetIgUrl = `https://www.instagram.com/p/${baseName}/`;
    const extension = path.extname(imgFile);
    const fileName = `${baseName}${extension}`;
    const imgPath = path.join(EXPORT_DIR, imgFile);

    const hasInDb = existingDbKeys.has(targetIgUrl);
    const hasInStorage = existingStorageFiles.has(fileName);

    if (hasInDb && hasInStorage) {
      alreadyExistInBoth++;
      continue;
    }

    // If missing in Storage, we need to upload it
    if (!hasInStorage) {
      uploadTasks.push({ baseName, fileName, imgPath });
      if (hasInDb) {
        onlyNeedUpload++;
      }
    }

    // If missing in DB, we need to insert it
    if (!hasInDb) {
      try {
        const caption = fs.readFileSync(path.join(EXPORT_DIR, txtFile), 'utf-8');
        const metadata = parseCaption(caption);
        const { data: { publicUrl } } = supabase.storage
          .from('cards')
          .getPublicUrl(`migrated/${fileName}`);

        cardsToInsert.push({
          title: metadata.title,
          description: caption,
          price: metadata.price,
          group_name: metadata.group,
          album_era: metadata.album_era,
          member_name: metadata.member,
          image_url: publicUrl,
          source: 'instagram',
          original_ig_url: targetIgUrl
        });
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : 'Unknown error';
        console.error(`❌ Failed processing metadata for ${baseName}:`, message);
      }
    }
  }

  console.log(`\n📊 Analysis Results:`);
  console.log(`⏭️ Already fully migrated (DB & Storage exist): ${alreadyExistInBoth}`);
  console.log(`📤 Images needing upload to Storage: ${uploadTasks.length} (out of which ${onlyNeedUpload} already exist in DB but missed in Storage)`);
  console.log(`➕ Cards needing database insertion: ${cardsToInsert.length}`);

  // 1. Upload missing images
  if (uploadTasks.length > 0) {
    console.log(`\n📤 Starting concurrent uploads for ${uploadTasks.length} images...`);
    let uploadedCount = 0;
    let failedCount = 0;

    await asyncPool(15, uploadTasks, async (task) => {
      try {
        const fileBuffer = fs.readFileSync(task.imgPath);
        const { error: uploadError } = await supabase.storage
          .from('cards')
          .upload(`migrated/${task.fileName}`, fileBuffer, { 
            contentType: 'image/jpeg',
            upsert: true 
          });
        if (uploadError) {
          console.error(`❌ Upload failed for ${task.fileName}:`, uploadError.message);
          failedCount++;
        } else {
          uploadedCount++;
          if (uploadedCount % 50 === 0 || uploadedCount === uploadTasks.length) {
            console.log(`   Uploaded ${uploadedCount}/${uploadTasks.length} images...`);
          }
        }
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        console.error(`❌ Exception uploading ${task.fileName}:`, message);
        failedCount++;
      }
    });
    console.log(`✅ Upload process completed. Success: ${uploadedCount}, Failed: ${failedCount}`);
  }

  // 2. Insert missing cards into database
  if (cardsToInsert.length > 0) {
    console.log(`\n🚀 Executing bulk database insertion...`);
    const chunkSize = 100;
    let insertedCount = 0;

    for (let i = 0; i < cardsToInsert.length; i += chunkSize) {
      const chunk = cardsToInsert.slice(i, i + chunkSize);
      const { error: insertErr } = await supabase
        .from('cards')
        .insert(chunk);

      if (insertErr) {
        console.error(`❌ Bulk insertion error for chunk starting at index ${i}:`, insertErr.message);
      } else {
        insertedCount += chunk.length;
        console.log(`   Inserted chunk ${i / chunkSize + 1} (${chunk.length} cards)`);
      }
    }
    console.log(`✅ Successfully inserted ${insertedCount} cards in bulk.`);
  }

  printSummary(cardsToInsert.length, 0, alreadyExistInBoth);
}

function printSummary(success: number, fail: number, skipped: number) {
  console.log(`\n🏁 Migration Complete!`);
  console.log(`📊 Success: ${success}`);
  console.log(`⏭️ Skipped (already exist): ${skipped}`);
  console.log(`❌ Failed: ${fail}`);
}

const GROUP_MAP: Record<string, string> = {
  'svt': 'Seventeen',
  'seventeen': 'Seventeen',
  '82major': '82major',
  '82m': '82major',
  'p1h': 'P1Harmony',
  'p1harmony': 'P1Harmony',
  'zb1': 'ZEROBASEONE',
  'zerobaseone': 'ZEROBASEONE',
  'lsf': 'LE SSERAFIM',
  'lesserafim': 'LE SSERAFIM',
  'adt': '&Team',
  '&team': '&Team',
  'amp': 'Ampers&one',
  'ampers&one': 'Ampers&one',
  'nct': 'NCT',
  'nctw': 'NCT Wish',
  'crv': 'CRAVITY',
  'cravity': 'CRAVITY',
  'bnd': 'BOYNEXTDOOR',
  'boynextdoor': 'BOYNEXTDOOR',
  'kfl': 'Kickflip',
  'kickflip': 'Kickflip',
  'cye': 'CYE',
  'ahf': 'ahof',
  'ahof': 'ahof',
  'xahf': 'ahof',
  'xlov': 'xlov',
  'nua': 'NouerA',
  'nouera': 'NouerA',
  'atz': 'Ateez',
  'ateez': 'Ateez',
  'cyn': 'Yena',
  'yena': 'Yena',
  'adp': 'Allday project',
  'allday': 'Allday project',
  'kik': 'kiiikiii',
  'kiiikiii': 'kiiikiii',
  'ehp': 'Enhypen',
  'enhypen': 'Enhypen',
  'xik': 'Xikers',
  'xikers': 'Xikers',
  'tws': 'TWS',
  'illit': 'Illit',
  'aespa': 'aespa',
  'h2h': 'aespa',
  'txt': 'TXT',
  'itzy': 'Itzy',
  'gmmtv': 'GMMTV',
  'evnne': 'Evnne',
  'rescene': 'RESCENE',
  'qwer': 'QWER',
  'kep1er': 'Kep1er',
  'kiss of life': 'Kiss of Life',
  'gidle': '(G)I-DLE',
  'i-dle': '(G)I-DLE',
  'exo': 'EXO',
  'kai': 'EXO',
  'chanyeol': 'EXO',
  'baekhyun': 'EXO',
  'riiz': 'Riize',
  'riize': 'Riize',
  'stray': 'Stray Kids',
  'skz': 'Stray Kids'
};

function parseCaption(caption: string) {
  const lines = caption.split('\n');
  const firstLine = lines[0]?.trim() || '';
  
  let price = 0;
  let group = '';
  let album_era = '';
  const member = '';

  // 1. Extract Price (handles "$35 set", "$24", etc.)
  const priceMatch = caption.match(/\$(\d+)(?:\s+set)?/i);
  if (priceMatch) {
    price = parseFloat(priceMatch[1]);
  }

  // 2. Parse Group, Album & Title
  if (firstLine.startsWith('#')) {
    const tokens = firstLine.split(/\s+/).filter(Boolean);
    const firstToken = tokens[0]; // e.g. "#meguronua" or "#riize"
    
    // Clean tag to extract abbreviation: remove leading '#', 'meguro', 'megurop', 'megurox'
    let rawTag = firstToken.substring(1).toLowerCase(); // remove '#'
    if (rawTag.startsWith('megurop')) {
      rawTag = rawTag.substring(7);
    } else if (rawTag.startsWith('megurox')) {
      rawTag = rawTag.substring(7);
    } else if (rawTag.startsWith('meguro')) {
      rawTag = rawTag.substring(6);
    }
    
    // Attempt to match Group from clean abbreviation
    if (GROUP_MAP[rawTag]) {
      group = GROUP_MAP[rawTag];
    }
    
    // If abbreviation not in map, check the other tokens for known group names
    if (!group) {
      for (const token of tokens) {
        const cleanedToken = token.replace(/[^a-zA-Z0-9&]/g, '').toLowerCase();
        if (GROUP_MAP[cleanedToken]) {
          group = GROUP_MAP[cleanedToken];
          break;
        }
      }
    }
    
    // If still not matched, fallback to the second token (legacy behavior)
    if (!group && tokens.length >= 2) {
      group = tokens[1];
    }
    
    // Filter out group tokens from the title/description parts
    const groupLower = group.toLowerCase();
    const groupWords = groupLower.split(/\s+/);
    
    const remainingTokens = tokens.slice(1).filter(t => {
      const cleanedT = t.toLowerCase().replace(/[^a-zA-Z0-9&]/g, '');
      if (t === firstToken) return false;
      if (cleanedT === rawTag) return false;
      if (groupWords.includes(cleanedT)) return false;
      return true;
    });

    if (remainingTokens.length >= 1) {
      album_era = remainingTokens[0];
    }
  }

  // 3. Extract Title (strip first token if it starts with #)
  let title = firstLine;
  if (firstLine.startsWith('#')) {
    const tokens = firstLine.split(/\s+/).filter(Boolean);
    if (tokens.length >= 2) {
      title = tokens.slice(1).join(' ');
    }
  }
  title = title.substring(0, 80).trim() || 'IG Post';

  return { title, price, group, album_era, member };
}

migrate();
